Rewrite of primary pointer section

patrickhlauke opened the pull request:
This generalises the treatment of mouse inputs to allow for the possibility of systems that support more than one simultaneous and independent mouse device, while at the same time clarifying that currently most (all?) operating systems/user agents don't have a concept of multiple mouse pointers. 

It future-proofs the spec while also reflecting reality.

Closes #49

patrickhlauke committed:
clarification why mouse is always primary

RByers commented:
This LGTM.  I also read over the compatibility mouse events section to ensure it was correct given these changes (definitely can't fire any mouse events for these weird non-primary mouse pointers), and it looks good.

RByers commented:
Just commenting on the precise wording though - I think we may want to make a couple tweaks...

RByers commented on index.html:
```diff
@@ -364,15 +364,17 @@
 				<div>When firing a pointer event, a pointer is considered primary if:
 					<ul>
-						<li>The pointer represents a mouse device.</li>
+						<li>The pointer represents a primary mouse input.</li>
```
Seems like we can simplify this whole section now that the mouse, touch and stylus paragraphs are mostly redundant.  See below.

RByers commented on index.html:
```diff
@@ -364,15 +364,17 @@
 					</ul>
 				</div>
 				<dl>
+					<dt>primary mouse input</dt><dd>A pointer representing mouse input is considered the <i>primary mouse input</i> if it dispatched any pointer events (such as <code>pointerdown</code> or <code>pointermove</code>) when no other active pointers representing a mouse input existed. However, see the note about <a href="#multiple-mouse-inputs">multiple mouse inputs</a>.</dd>
```
I think this (and the existing touch/pen sections below when considering hover-capable pointers) isn't quite right.  Eg. consider the case of two mice (or pens) both hovering without either being active.  One needs to be primary, right?  What happens in Edge today when a pen first becomes visibly by the digitizer - does it have `isPrimary=true` from the first `pointermove`, or only after the first `pointerdown` event?

I think the essential properties are that:
- At any given time exactly one pointer of each type is considered primary
- When the first pointer of a given type becomes active, that pointer also becomes the primary pointer

patrickhlauke commented:
@RByers 

> What happens in Edge today when a pen first becomes visibly by the digitizer - does it have isPrimary=true from the first pointermove, or only after the first pointerdown event?

testing on Surface 3 with Pen and Edge, the pen is primary as soon as it's recognized/hovering. so quite right, i think this isn't correct - but that's a separate issue (which i'll file later today)

patrickhlauke commented:
> When the first pointer of a given type becomes active, that pointer also becomes the primary pointer

agree - perhaps with link/pointer to the glossary definition of "active pointer" - this is the simplest amendment (which then can help simplify the entire prose for this section, i may actually work on this as one combined PR, but open a separate issue for clarity)

patrickhlauke committed:
Simplification and correction of primary pointer section

Due to the addition of multi-mouse scenario, the explanation of what a
primary pointer is can be condensed/simplified, rather than being split
into "primary mouse"/"primary touch"/"primary pointer".

This change also corrects the previously erroneous description of when
touch/pen pointers become primary - as hovered pen (and presumably, if
supported by hardware, hovered touch) is also primary, it's not really
related to `pointermove` but rather to which active pointer for that
type came first.

patrickhlauke commented:
Pushed a separate commit to address both #54 and #49 - drastically simplifies the definition and brings it in line with reality of implementation.

/cc @jacobrossi @RByers

mustaqahmed commented on index.html:
```diff
@@ -359,28 +359,17 @@
-			</section>
+			<div><p>In a multi-pointer (e.g. multi-touch) scenario, the <a href="#widl-PointerEvent-isPrimary"><code>isPrimary</code></a> property is used to identify a master pointer amongst the set of <a data-lt="active pointer">active pointers</a> for each pointer type.</p></div>
+			<ul>
+				<li>At any given time, there can only ever be exactly one primary pointer for each pointer type.</li>
```
Does
  "...there can be _at most one_ primary pointer..."
make it clearer? The phrase "can ever be exactly one" seems to exclude the following cases of no primary "touch" pointer:
- when there's no active touches, and
- when two fingers touch the screen in sequence, then the first one is removed.

patrickhlauke replied:
ah, good call. yes, that would make it clearer and more correct. i'll push an update

mustaqahmed commented on index.html:
```diff
@@ -359,28 +359,17 @@
+				<li>The first pointer to become active for a particular pointer type (e.g. the first finger to touch the screen in a multi-touch interaction) becomes the primary pointer for that pointer type.</li>
+				<li>Only a primary pointer will produce <a>compatibility mouse events</a>.</li>
+			</ul>
+			<div class="note">Authors who desire single-pointer interaction can achieve this by ignoring non-primary pointers (however, see the note below on <a href="#multiple-primary-pointers">multiple primary pointers</a>).</p></div>
```
Nit: Is there a way to avoid 5 Notes in a row? E.g. by combining 1 & 2, and/or dropping 3?

patrickhlauke replied:
3 could be added part of the second bullet above ("Only a primary pointer will product..."). that should lighten this up a bit. 1 & 2 are not directly related (one is about multiple pointers of the same type, the other about different pointer types and their primary pointers), so i'd leave them separate.

mustaqahmed commented on index.html:
```diff
@@ -359,28 +359,16 @@
+			<ul>
+				<li>At any given time, there can only ever be at most one primary pointer for each pointer type.</li>
+				<li>The first pointer to become active for a particular pointer type (e.g. the first finger to touch the screen in a multi-touch interaction) becomes the primary pointer for that pointer type.</li>
+				<li>Only a primary pointer will produce <a>compatibility mouse events</a>. In the case where there are multiple <a href="#the-primary-pointer">primary pointers</a>, these pointers will all produce <a>compatibility mouse events</a></li>
```
We are missing a '.' at the end.

patrickhlauke replied:
good spot. fixed

mustaqahmed commented:
LGTM, thanks Patrick.

RByers commented:
LGTM, thanks!

RByers commented:
I squashed your two commits and merged them.  I hope that's what you intended (otherwise I can easily undo it).

patrickhlauke commented:
Yes i was relying on GitHub's new autosquashing feature :)
